// nebula.gl/dist/lib/layers/texts-layer.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _deck = require('deck.gl');

var _nebulaLayer = require('../nebula-layer');

var _nebulaLayer2 = _interopRequireDefault(_nebulaLayer);

var _utils = require('../utils');

var _deckCache2 = require('../deck-renderer/deck-cache');

var _deckCache3 = _interopRequireDefault(_deckCache2);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var TextsLayer = function (_NebulaLayer) {
  _inherits(TextsLayer, _NebulaLayer);

  function TextsLayer(config) {
    _classCallCheck(this, TextsLayer);

    var _this = _possibleConstructorReturn(this, (TextsLayer.__proto__ || Object.getPrototypeOf(TextsLayer)).call(this, config));

    _this.deckCache = new _deckCache3.default(config.getData, function (data) {
      return config.toNebulaFeature(data);
    });
    return _this;
  }

  _createClass(TextsLayer, [{
    key: 'render',
    value: function render(_ref) {
      var nebula = _ref.nebula;

      var defaultColor = [0x0, 0x0, 0x0, 0xff];
      var _deckCache = this.deckCache,
          objects = _deckCache.objects,
          updateTrigger = _deckCache.updateTrigger;
      var zoom = nebula.props.viewport.zoom;


      return new _deck.TextLayer({
        id: 'texts-' + this.id,
        data: objects,
        opacity: 1,
        fp64: false,
        pickable: false,

        getText: function getText(nf) {
          return nf.style.text;
        },
        getPosition: function getPosition(nf) {
          return nebula.projector.coordsToLngLatOffset(nf.geoJson.geometry.coordinates);
        },
        getColor: function getColor(nf) {
          return (0, _utils.toDeckColor)(nf.style.fillColor) || defaultColor;
        },

        // TODO: layer should offer option to scale with zoom
        sizeScale: 1 / Math.pow(2, 20 - zoom),

        coordinateSystem: _deck.COORDINATE_SYSTEM.LNGLAT_OFFSETS,
        coordinateOrigin: nebula.projector.lngLat,
        updateTriggers: { all: updateTrigger + '_' + nebula.projector.lngLat },

        nebulaLayer: this
      });
    }
  }]);

  return TextsLayer;
}(_nebulaLayer2.default);

exports.default = TextsLayer;
