// nebula.gl/dist-es6/lib/layers/junctions-layer.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _deck = require('deck.gl');

var _junctionScatterplotLayer = require('../deck-renderer/junction-scatterplot-layer');

var _junctionScatterplotLayer2 = _interopRequireDefault(_junctionScatterplotLayer);

var _nebulaLayer = require('../nebula-layer');

var _nebulaLayer2 = _interopRequireDefault(_nebulaLayer);

var _utils = require('../utils');

var _deckCache2 = require('../deck-renderer/deck-cache');

var _deckCache3 = _interopRequireDefault(_deckCache2);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var JunctionsLayer = function (_NebulaLayer) {
  _inherits(JunctionsLayer, _NebulaLayer);

  function JunctionsLayer(config) {
    _classCallCheck(this, JunctionsLayer);

    var _this = _possibleConstructorReturn(this, (JunctionsLayer.__proto__ || Object.getPrototypeOf(JunctionsLayer)).call(this, config));

    _this.deckCache = new _deckCache3.default(config.getData, function (data) {
      return config.toNebulaFeature(data);
    });
    _this.enablePicking = true;
    return _this;
  }

  _createClass(JunctionsLayer, [{
    key: 'render',
    value: function render(_ref) {
      var nebula = _ref.nebula;

      var defaultColor = [0x0, 0x0, 0x0, 0xff];
      var _deckCache = this.deckCache,
          objects = _deckCache.objects,
          updateTrigger = _deckCache.updateTrigger;


      return new _junctionScatterplotLayer2.default({
        id: 'junctions-' + this.id,
        data: objects,
        opacity: 1,
        fp64: false,
        pickable: true,
        getPosition: function getPosition(nf) {
          return nebula.projector.coordsToLngLatOffset(nf.geoJson.geometry.coordinates);
        },
        getFillColor: function getFillColor(nf) {
          return (0, _utils.toDeckColor)(nf.style.fillColor) || defaultColor;
        },
        getStrokeColor: function getStrokeColor(nf) {
          return (0, _utils.toDeckColor)(nf.style.outlineColor) || (0, _utils.toDeckColor)(nf.style.fillColor) || defaultColor;
        },
        getRadius: function getRadius(nf) {
          return nf.style.pointRadiusMeters + nf.style.outlineRadiusMeters || 1;
        },
        getInnerRadius: function getInnerRadius(nf) {
          return nf.style.pointRadiusMeters || 0.5;
        },
        parameters: {
          depthTest: false,
          blend: false
        },

        coordinateSystem: _deck.COORDINATE_SYSTEM.LNGLAT_OFFSETS,
        coordinateOrigin: nebula.projector.lngLat,
        updateTriggers: { all: updateTrigger + '_' + nebula.projector.lngLat },

        nebulaLayer: this
      });
    }
  }]);

  return JunctionsLayer;
}(_nebulaLayer2.default);

exports.default = JunctionsLayer;
