// fusion-plugin-styletron-react/dist/index.es.js
import React from 'react';
import path from 'path';
import fs from 'fs';
import { createPlugin, dangerouslySetHTML } from 'fusion-core';
import { Provider, styled, withStyle, withStyleDeep, withTransform, withWrapper } from 'styletron-react';
export { styled, withStyle, withStyleDeep, withTransform, withWrapper } from 'styletron-react';
import { Server } from 'styletron-engine-atomic';

/** Copyright (c) 2018 Uber Technologies, Inc.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * 
 */
/**
 * Provides styletron instance via old context API
 */

class LegacyStyletronProvider extends React.Component {
  getChildContext() {
    return {
      styletron: this.styletron
    };
  } // $FlowFixMe


  constructor(props, context) {
    super(props, context);
    this.styletron = props.value;
  }

  render() {
    return React.Children.only(this.props.children);
  }

}

LegacyStyletronProvider.childContextTypes = {
  styletron: noop
};
function noop() {}

/** Copyright (c) 2018 Uber Technologies, Inc.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * 
 */
// $FlowFixMe
const injectDeclarationCompatMixin = Base => class extends Base {
  // $FlowFixMe
  injectDeclaration({
    prop,
    val,
    media,
    pseudo
  }) {
    let style = {
      [prop]: val
    };

    if (pseudo) {
      style = {
        [pseudo]: style
      };
    }

    if (media) {
      style = {
        [media]: style
      };
    }

    return this.renderStyle(style);
  }

};

/** Copyright (c) 2018 Uber Technologies, Inc.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * 
 */
const workerRoute = '/__styletron_debugger_worker.js';
const wasmRoute = '/__styletron_debugger_mappings.wasm';

/** Copyright (c) 2018 Uber Technologies, Inc.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * 
 */

/* eslint-env node */
let workerPath;
let wasmPath;

if (process.env.NODE_ENV !== "production" && true) {
  const base = path.dirname(require.resolve('css-to-js-sourcemap-worker'));
  workerPath = path.resolve(base, 'worker.js');
  wasmPath = path.resolve(base, 'mappings.wasm');
}

const StyletronCompat = injectDeclarationCompatMixin(Server);
const plugin = true && createPlugin({
  middleware: () => (ctx, next) => {
    if (process.env.NODE_ENV !== "production") {
      if (ctx.url === workerRoute) {
        ctx.body = fs.createReadStream(workerPath);
        return next();
      }

      if (ctx.url === wasmRoute) {
        ctx.body = fs.createReadStream(wasmPath);
        return next();
      }
    }

    if (ctx.element) {
      const engine = new StyletronCompat();
      ctx.element = React.createElement(Provider, {
        value: engine
      }, React.createElement(LegacyStyletronProvider, {
        value: engine
      }, ctx.element));
      return next().then(() => {
        const stylesForHead = engine.getStylesheetsHtml();
        ctx.template.head.push(dangerouslySetHTML(stylesForHead));
      });
    } else {
      return next();
    }
  }
});

/** Copyright (c) 2018 Uber Technologies, Inc.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * 
 */

/* eslint-env browser */

/* global module */

/** Copyright (c) 2018 Uber Technologies, Inc.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * 
 */

/* eslint-env browser */
var index = plugin;

export default index;
