// nebula.gl/dist-es6/lib/overlays/html-overlay-item.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _slicedToArray = function () { function sliceIterator(arr, i) { var _arr = []; var _n = true; var _d = false; var _e = undefined; try { for (var _i = arr[Symbol.iterator](), _s; !(_n = (_s = _i.next()).done); _n = true) { _arr.push(_s.value); if (i && _arr.length === i) break; } } catch (err) { _d = true; _e = err; } finally { try { if (!_n && _i["return"]) _i["return"](); } finally { if (_d) throw _e; } } return _arr; } return function (arr, i) { if (Array.isArray(arr)) { return arr; } else if (Symbol.iterator in Object(arr)) { return sliceIterator(arr, i); } else { throw new TypeError("Invalid attempt to destructure non-iterable instance"); } }; }();

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var styles = {
  item: {
    position: 'absolute',
    userSelect: 'none'
  }
};

var HtmlOverlayItem = function (_Component) {
  _inherits(HtmlOverlayItem, _Component);

  function HtmlOverlayItem() {
    _classCallCheck(this, HtmlOverlayItem);

    return _possibleConstructorReturn(this, (HtmlOverlayItem.__proto__ || Object.getPrototypeOf(HtmlOverlayItem)).apply(this, arguments));
  }

  _createClass(HtmlOverlayItem, [{
    key: 'render',
    value: function render() {
      var _props = this.props,
          children = _props.children,
          coordinates = _props.coordinates,
          style = _props.style,
          props = _objectWithoutProperties(_props, ['children', 'coordinates', 'style']);

      var _context$parent$getCo = this.context.parent.getCoords(coordinates),
          _context$parent$getCo2 = _slicedToArray(_context$parent$getCo, 2),
          x = _context$parent$getCo2[0],
          y = _context$parent$getCo2[1];

      if (this.context.parent.inView([x, y])) return _react2.default.createElement(
        'div',
        _extends({ style: _extends({}, styles.item, style, { left: x, top: y }) }, props),
        children
      );

      return null;
    }
  }]);

  return HtmlOverlayItem;
}(_react.Component);

HtmlOverlayItem.contextTypes = {
  parent: _propTypes2.default.object
};
exports.default = HtmlOverlayItem;
