// nebula.gl/dist-es6/lib/overlays/html-overlay.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _slicedToArray = function () { function sliceIterator(arr, i) { var _arr = []; var _n = true; var _d = false; var _e = undefined; try { for (var _i = arr[Symbol.iterator](), _s; !(_n = (_s = _i.next()).done); _n = true) { _arr.push(_s.value); if (i && _arr.length === i) break; } } catch (err) { _d = true; _e = err; } finally { try { if (!_n && _i["return"]) _i["return"](); } finally { if (_d) throw _e; } } return _arr; } return function (arr, i) { if (Array.isArray(arr)) { return arr; } else if (Symbol.iterator in Object(arr)) { return sliceIterator(arr, i); } else { throw new TypeError("Invalid attempt to destructure non-iterable instance"); } }; }();

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _nebulaOverlay = require('./nebula-overlay');

var _nebulaOverlay2 = _interopRequireDefault(_nebulaOverlay);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var styles = {
  mainContainer: {
    width: '100%',
    height: '100%',
    position: 'absolute',
    pointerEvents: 'none'
  }
};

var HtmlOverlay = function (_NebulaOverlay) {
  _inherits(HtmlOverlay, _NebulaOverlay);

  function HtmlOverlay() {
    _classCallCheck(this, HtmlOverlay);

    return _possibleConstructorReturn(this, (HtmlOverlay.__proto__ || Object.getPrototypeOf(HtmlOverlay)).apply(this, arguments));
  }

  _createClass(HtmlOverlay, [{
    key: 'getChildContext',
    value: function getChildContext() {
      return { parent: this };
    }

    // override this to provide your items

  }, {
    key: 'getItems',
    value: function getItems() {
      throw new Error('Please provide your own items');
    }
  }, {
    key: 'getCoords',
    value: function getCoords(coordinates) {
      var pos = this.context.wmViewport.project(coordinates);
      if (!pos) return [-1, -1];
      return pos;
    }
  }, {
    key: 'inView',
    value: function inView(_ref) {
      var _ref2 = _slicedToArray(_ref, 2),
          x = _ref2[0],
          y = _ref2[1];

      var _context$viewport = this.context.viewport,
          width = _context$viewport.width,
          height = _context$viewport.height;

      return !(x < 0 || y < 0 || x > width || y > height);
    }
  }, {
    key: 'scaleWithZoom',
    value: function scaleWithZoom(n) {
      var zoom = this.context.viewport.zoom;

      return n / Math.pow(2, 20 - zoom);
    }
  }, {
    key: 'breakpointWithZoom',
    value: function breakpointWithZoom(threshold, a, b) {
      var zoom = this.context.viewport.zoom;

      return zoom > threshold ? a : b;
    }
  }, {
    key: 'getViewport',
    value: function getViewport() {
      return this.context.viewport;
    }
  }, {
    key: 'getZoom',
    value: function getZoom() {
      return this.context.viewport.zoom;
    }
  }, {
    key: 'render',
    value: function render() {
      return _react2.default.createElement(
        'div',
        { style: styles.mainContainer },
        this.getItems()
      );
    }
  }]);

  return HtmlOverlay;
}(_nebulaOverlay2.default);

HtmlOverlay.childContextTypes = {
  parent: _propTypes2.default.object
};
exports.default = HtmlOverlay;
