// nebula.gl/dist-es6/lib/overlays/html-cluster-overlay.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _helpers = require('@turf/helpers');

var _supercluster = require('supercluster');

var _supercluster2 = _interopRequireDefault(_supercluster);

var _htmlOverlay = require('./html-overlay');

var _htmlOverlay2 = _interopRequireDefault(_htmlOverlay);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var HtmlClusterOverlay = function (_HtmlOverlay) {
  _inherits(HtmlClusterOverlay, _HtmlOverlay);

  function HtmlClusterOverlay() {
    var _ref;

    var _temp, _this, _ret;

    _classCallCheck(this, HtmlClusterOverlay);

    for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    return _ret = (_temp = (_this = _possibleConstructorReturn(this, (_ref = HtmlClusterOverlay.__proto__ || Object.getPrototypeOf(HtmlClusterOverlay)).call.apply(_ref, [this].concat(args))), _this), _this._lastObjects = null, _temp), _possibleConstructorReturn(_this, _ret);
  }

  _createClass(HtmlClusterOverlay, [{
    key: 'getItems',
    value: function getItems() {
      var _this2 = this;

      // supercluster().load() is expensive and we want to run it only
      // when necessary and not for every frame.
      var newObjects = this.getAllObjects();
      if (newObjects !== this._lastObjects) {
        this._superCluster = (0, _supercluster2.default)({ maxZoom: 20 });
        this._superCluster.load(newObjects.map(function (object) {
          return (0, _helpers.point)(_this2.getObjectCoordinates(object), { object: object });
        }));
        this._lastObjects = newObjects;
      }

      var clusters = this._superCluster.getClusters([-180, -90, 180, 90], Math.round(this.getZoom()));

      return clusters.map(function (_ref2) {
        var coordinates = _ref2.geometry.coordinates,
            _ref2$properties = _ref2.properties,
            cluster = _ref2$properties.cluster,
            pointCount = _ref2$properties.point_count,
            clusterId = _ref2$properties.cluster_id,
            object = _ref2$properties.object;
        return cluster ? _this2.renderCluster(coordinates, clusterId, pointCount) : _this2.renderObject(coordinates, object);
      });
    }
  }, {
    key: 'getClusterObjects',
    value: function getClusterObjects(clusterId) {
      return this._superCluster.getLeaves(clusterId, Math.round(this.getZoom()), Infinity).map(function (object) {
        return object.properties.object;
      });
    }

    // Override to provide items that need clustering.
    // If the items have not changed please provide the same array to avoid
    // regeneration of the cluster which causes performance issues.

  }, {
    key: 'getAllObjects',
    value: function getAllObjects() {
      return [];
    }

    // override to provide coordinates for each object of getAllObjects()

  }, {
    key: 'getObjectCoordinates',
    value: function getObjectCoordinates(obj) {
      return [0, 0];
    }

    // override to return an HtmlOverlayItem

  }, {
    key: 'renderObject',
    value: function renderObject(coordinates, obj) {
      return null;
    }

    // override to return an HtmlOverlayItem
    // use getClusterObjects() to get cluster contents

  }, {
    key: 'renderCluster',
    value: function renderCluster(coordinates, clusterId, pointCount) {
      return null;
    }
  }]);

  return HtmlClusterOverlay;
}(_htmlOverlay2.default);

exports.default = HtmlClusterOverlay;
