// nebula.gl/dist-es6/lib/deck-renderer/junction-scatterplot-layer.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _deck = require('deck.gl');

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var JunctionScatterplotLayer = function (_CompositeLayer) {
  _inherits(JunctionScatterplotLayer, _CompositeLayer);

  function JunctionScatterplotLayer() {
    _classCallCheck(this, JunctionScatterplotLayer);

    return _possibleConstructorReturn(this, (JunctionScatterplotLayer.__proto__ || Object.getPrototypeOf(JunctionScatterplotLayer)).apply(this, arguments));
  }

  _createClass(JunctionScatterplotLayer, [{
    key: 'renderLayers',
    value: function renderLayers() {
      var _props = this.props,
          id = _props.id,
          getFillColor = _props.getFillColor,
          getStrokeColor = _props.getStrokeColor,
          getInnerRadius = _props.getInnerRadius,
          updateTriggers = _props.updateTriggers;

      // data needs to be passed explicitly after deck.gl 5.3

      return [
      // the full circles
      new _deck.ScatterplotLayer(_extends({}, this.props, {
        id: id + '-full',
        data: this.props.data,
        getColor: getStrokeColor,
        updateTriggers: _extends({}, updateTriggers, {
          getColor: updateTriggers.getStrokeColor
        })
      })),
      // the inner part
      new _deck.ScatterplotLayer(_extends({}, this.props, {
        id: id + '-inner',
        data: this.props.data,
        getColor: getFillColor,
        getRadius: getInnerRadius,
        pickable: false,
        updateTriggers: _extends({}, updateTriggers, {
          getColor: updateTriggers.getFillColor,
          getRadius: updateTriggers.getInnerRadius
        })
      }))];
    }
  }]);

  return JunctionScatterplotLayer;
}(_deck.CompositeLayer);

JunctionScatterplotLayer.layerName = 'JunctionScatterplotLayer';
JunctionScatterplotLayer.defaultProps = _extends({}, _deck.ScatterplotLayer.defaultProps, {
  getFillColor: function getFillColor(d) {
    return [0, 0, 0, 255];
  },
  getStrokeColor: function getStrokeColor(d) {
    return [255, 255, 255, 255];
  },
  getInnerRadius: function getInnerRadius(d) {
    return 1;
  }
});
exports.default = JunctionScatterplotLayer;
