// nebula.gl/dist/lib/overlays/html-tooltip-overlay.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _window = require('global/window');

var _window2 = _interopRequireDefault(_window);

var _htmlOverlay = require('./html-overlay');

var _htmlOverlay2 = _interopRequireDefault(_htmlOverlay);

var _htmlOverlayItem = require('./html-overlay-item');

var _htmlOverlayItem2 = _interopRequireDefault(_htmlOverlayItem);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var styles = {
  tooltip: {
    transform: 'translate(-50%,-100%)',
    backgroundColor: 'rgba(0, 0, 0, 0.3)',
    padding: '4px 8px',
    borderRadius: 8,
    color: 'white'
  }
};

var SHOW_TOOLTIP_TIMEOUT = 250;

var HtmlTooltipOverlay = function (_HtmlOverlay) {
  _inherits(HtmlTooltipOverlay, _HtmlOverlay);

  function HtmlTooltipOverlay(props) {
    _classCallCheck(this, HtmlTooltipOverlay);

    var _this = _possibleConstructorReturn(this, (HtmlTooltipOverlay.__proto__ || Object.getPrototypeOf(HtmlTooltipOverlay)).call(this, props));

    _this.timeoutID = null;

    _this.state = { visible: false, pickingInfo: null };
    return _this;
  }

  _createClass(HtmlTooltipOverlay, [{
    key: 'componentWillMount',
    value: function componentWillMount() {
      var _this2 = this;

      this.context.nebula.queryObjectEvents.on('pick', function (_ref) {
        var event = _ref.event,
            pickingInfo = _ref.pickingInfo;

        if (_this2.timeoutID !== null) {
          _window2.default.clearTimeout(_this2.timeoutID);
        }
        _this2.timeoutID = null;

        if (pickingInfo && _this2._getTooltip(pickingInfo)) {
          _this2.timeoutID = _window2.default.setTimeout(function () {
            _this2.setState({ visible: true, pickingInfo: pickingInfo });
          }, SHOW_TOOLTIP_TIMEOUT);
        } else {
          _this2.setState({ visible: false });
        }
      });
    }
  }, {
    key: '_getTooltip',
    value: function _getTooltip(pickingInfo) {
      return pickingInfo.object.style.tooltip;
    }
  }, {
    key: '_makeOverlay',
    value: function _makeOverlay() {
      var pickingInfo = this.state.pickingInfo;


      if (pickingInfo) {
        return _react2.default.createElement(
          _htmlOverlayItem2.default,
          { key: 0, coordinates: pickingInfo.lngLat, style: styles.tooltip },
          this._getTooltip(pickingInfo)
        );
      }

      return null;
    }
  }, {
    key: 'getItems',
    value: function getItems() {
      if (this.state.visible) {
        return [this._makeOverlay()];
      }

      return [];
    }
  }]);

  return HtmlTooltipOverlay;
}(_htmlOverlay2.default);

exports.default = HtmlTooltipOverlay;
