// nebula.gl/dist-es6/lib/layer-mouse-event.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

var LayerMouseEvent = function () {
  // browser event

  // internal nebula info about the object
  function LayerMouseEvent(nativeEvent, _ref) {
    var data = _ref.data,
        groundPoint = _ref.groundPoint,
        nebula = _ref.nebula,
        metadata = _ref.metadata;

    _classCallCheck(this, LayerMouseEvent);

    this.nativeEvent = nativeEvent;

    this.data = data;
    this.groundPoint = groundPoint;
    this.nebula = nebula;
    this.metadata = metadata;
  }
  // reference to nebula

  // the mouse [lng,lat] raycasted onto the ground

  // original item that this event is related to


  _createClass(LayerMouseEvent, [{
    key: 'stopPropagation',
    value: function stopPropagation() {
      this.nativeEvent.stopPropagation();
      this.canceled = true;
    }
  }]);

  return LayerMouseEvent;
}();

exports.default = LayerMouseEvent;
