// nebula.gl/dist-es6/lib/nebula-react.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _deck = require('deck.gl');

var _deck2 = _interopRequireDefault(_deck);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _nebula = require('./nebula');

var _nebula2 = _interopRequireDefault(_nebula);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var styles = {
  canvasContainer: {
    pointerEvents: 'none',
    position: 'absolute',
    left: 0,
    right: 0,
    top: 0,
    bottom: 0
  }
};

var NebulaReact = function (_Component) {
  _inherits(NebulaReact, _Component);

  function NebulaReact() {
    _classCallCheck(this, NebulaReact);

    return _possibleConstructorReturn(this, (NebulaReact.__proto__ || Object.getPrototypeOf(NebulaReact)).apply(this, arguments));
  }

  _createClass(NebulaReact, [{
    key: 'getChildContext',
    value: function getChildContext() {
      var nebula = this.nebula;
      var viewport = nebula.props.viewport,
          wmViewport = nebula.wmViewport;

      return { viewport: viewport, wmViewport: wmViewport, nebula: nebula };
    }
  }, {
    key: 'componentWillMount',
    value: function componentWillMount() {
      var _this2 = this;

      this.nebula = new _nebula2.default();
      this.nebula.init(this.props);
      this.nebula.forceUpdate = function () {
        return _this2.forceUpdate();
      };
    }
  }, {
    key: 'componentWillReceiveProps',
    value: function componentWillReceiveProps(props) {
      this.nebula.updateProps(props);
    }
  }, {
    key: 'updateAllDeckObjects',
    value: function updateAllDeckObjects() {
      this.nebula.updateAllDeckObjects();
    }
  }, {
    key: 'updateDeckObjectsByIds',
    value: function updateDeckObjectsByIds(ids) {
      this.nebula.updateDeckObjectsByIds(ids);
    }
  }, {
    key: 'render',
    value: function render() {
      var _this3 = this;

      var _props = this.props,
          viewport = _props.viewport,
          children = _props.children;
      var width = viewport.width,
          height = viewport.height;


      return _react2.default.createElement(
        'div',
        { style: styles.canvasContainer, ref: function ref(div) {
            return _this3.mainContainer = div;
          } },
        _react2.default.createElement(_deck2.default, {
          ref: function ref(deckgl) {
            return _this3.nebula.setDeck(deckgl);
          },
          width: width,
          height: height,
          viewports: [this.nebula.wmViewport],
          layers: this.nebula.getRenderedLayers()
        }),
        children
      );
    }
  }]);

  return NebulaReact;
}(_react.Component);

NebulaReact.childContextTypes = {
  viewport: _propTypes2.default.object,
  wmViewport: _propTypes2.default.object,
  nebula: _propTypes2.default.object
};
exports.default = NebulaReact;
